plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -1574,16 +1576,67 @@ public void setColumnType(ConnectorSession session, ConnectorTableHandle tableHa
         verify(column.isBaseColumn(), "Cannot change nested field types");
 
         Table icebergTable = catalog.loadTable(session, table.getSchemaTableName());
+        Type sourceType = icebergTable.schema().findType(column.getName());
+        Type newType = toIcebergType(type);
         try {
-            icebergTable.updateSchema()
-                    .updateColumn(column.getName(), toIcebergType(type).asPrimitiveType())
+            buildUpdateSchema(column.getName(), sourceType, newType, icebergTable.updateSchema())
                     .commit();
         }
         catch (RuntimeException e) {
             throw new TrinoException(ICEBERG_COMMIT_ERROR, "Failed to set column type: " + firstNonNull(e.getMessage(), e), e);
         }
     }
 
+    private static UpdateSchema buildUpdateSchema(String name, Type sourceType, Type newType, UpdateSchema schema)
```

Comment:
Nit: I see that returning the UpdateSchema makes calling `commit` easier above, but personally find this easier to read.
```suggestion
    private static void buildUpdateSchema(String name, Type sourceType, Type newType, UpdateSchema schemaUpdate)
```